core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,483 @@
+    write(mSingleCharWrite);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (b == null || len == 0) {
+      return;
+    }
+    validateWriteArgs(b, off, len);
+    if (mFile == null) {
+      initNewFile();
+    }
+    if (mPartitionOffset + len <= mPartitionSize) {
+      mLocalOutputStream.write(b, off, len);
+      mPartitionOffset += len;
+    } else {
+      int firstLen = (int) (mPartitionSize - mPartitionOffset);
+      mLocalOutputStream.write(b, off, firstLen);
+      mPartitionOffset += firstLen;
+      uploadPart();
+      write(b, off + firstLen, len - firstLen);
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    if (!isMultiPartUploadInitialized()) {
+      return;
+    }
+    // We try to minimize the time use to close()
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    if (mLocalOutputStream != null) {
+      mLocalOutputStream.flush();
+    }
+    if (mPartitionOffset > UPLOAD_THRESHOLD) {
+      uploadPart();
+    }
+    waitForAllPartsUpload();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+
+    // Set the closed flag, we never retry close() even if exception occurs
+    mClosed = true;
+
+    // Multi-part upload has not been initialized
+    if (!isMultiPartUploadInitialized()) {
+      if (mFile == null) {
+        LOG.debug("Streaming upload output stream closed without uploading any data.");
+        putEmptyObject();
+      } else {
+        try {
+          mLocalOutputStream.close();
+          putObjectDirectly();
+        } finally {
+          if (!mFile.delete()) {
+            LOG.error("Failed to delete temporary file @ {}", mFile.getPath());
+          }
+        }
+      }
+      return;
+    }
+
+    try {
+      if (mFile != null) {
+        mLocalOutputStream.close();
+        int partNumber = mPartNumber.getAndIncrement();
+        uploadPart(mFile, partNumber, true);
+      }
+
+      waitForAllPartsUpload();
+      completeMultiPartUpload();
+    } catch (Exception e) {
+      LOG.error("Failed to upload {}", mKey, e);
+      throw new IOException(e);
+    }
+  }
+
+  private void putObjectDirectly() throws IOException {
+    final String md5;
+    if (mHash != null) {
+      md5 = Base64.encodeBase64String(mHash.digest());
+    } else {
+      md5 = null;
+    }
+    IOException lastException;
+    do {
+      try {
+        putObject(mKey, mFile, md5);
+        return;
+      } catch (IOException e) {
+        lastException = e;
+      }
+    } while (mRetryPolicy.attempt());
+    throw lastException;
+  }
+
+  private void putEmptyObject() throws IOException {
+    IOException lastException;
+    do {
```

Comment:
similar to other methods, try to see if we can find something in RetryUtils

Reply:
fix